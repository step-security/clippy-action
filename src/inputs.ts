/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023-2024 Noel Towa <cutie@floofy.dev>
 * Copyright (c) 2025 StepSecurity
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type InputOptions, getInput, error, warning } from '@actions/core';

export interface Inputs {
    'working-directory'?: string;
    'all-features': boolean;
    'github-token': string;
    'check-args': string[];
    forbid: string[];
    allow: string[];
    deny: string[];
    args: string[];
    warn: string[];
}

const truthy = new Set(['true', 'True', 'TRUE']);
const falsy = new Set(['false', 'False', 'FALSE']);

const getBooleanInput = (name: string, options: InputOptions) => {
    const value = getInput(name, options);
    if (!value) return false;
    if (truthy.has(value)) return true;
    if (falsy.has(value)) return false;

    throw new Error(
        `Value of key [${name}] didn't meet the Yaml 1.2 "Core Schema" specification (received: [${value}])`
    );
};

function getArrayInput(input: string, opts: { sep: string }) {
    return getInput(input, { trimWhitespace: true })
        .split(opts.sep)
        .filter(Boolean)
        .map((s) => s.trim());
}

export function getInputs(): Inputs | null {
    const wd = getInput('working-directory', { trimWhitespace: true });
    const token = getInput('token', { trimWhitespace: true });
    if (token === '') {
        error('Missing required input: `token`');
        return null;
    }

    const allow = getArrayInput('allow', { sep: ',' });
    if (allow.length) {
        warning('The `allow` input is deprecated in v1.4 and can be used within the `check-args` input');
    }

    const deny = getArrayInput('deny', { sep: ',' });
    if (deny.length) {
        warning('The `deny` input is deprecated in v1.4 and can be used within the `check-args` input');
    }

    const forbid = getArrayInput('forbid', { sep: ',' });
    if (forbid.length) {
        warning('The `forbid` input is deprecated in v1.4 and can be used within the `check-args` input');
    }

    const warn = getArrayInput('warn', { sep: ',' });
    if (warn.length) {
        warning('The `warn` input is deprecated in v1.4 and can be used within the `check-args` input');
    }

    const args = getArrayInput('args', { sep: ' ' });
    const checkArgs = getArrayInput('check-args', { sep: ' ' });

    return {
        'working-directory': wd,
        'github-token': token,
        'all-features': getBooleanInput('all-features', { trimWhitespace: true }),
        'check-args': checkArgs,
        forbid,
        allow,
        warn,
        deny,
        args
    } satisfies Inputs;
}
