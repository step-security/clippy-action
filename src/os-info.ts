/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023-2024 Noel Towa <cutie@floofy.dev>
 * Copyright (c) 2025 StepSecurity
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export function os() {
    switch (process.platform) {
        case 'aix':
            return 'IBM AIX';

        case 'android':
            return 'Android';

        case 'darwin':
            return 'macOS';

        case 'freebsd':
            return 'FreeBSD';

        case 'haiku':
            return 'HaikuOS';

        case 'linux':
            return 'Linux';

        case 'openbsd':
            return 'OpenBSD';

        case 'sunos':
            return 'SunOS';

        case 'win32':
        case 'cygwin':
            return 'Windows';

        case 'netbsd':
            return 'NetBSD';

        default:
            return 'Unknown';
    }
}

export function arch() {
    switch (process.arch) {
        case 'arm':
            return 'ARM';

        case 'arm64':
            return 'ARM64';

        case 'ia32':
            return 'x86';

        case 'x64':
            return 'x86_64';

        default:
            return 'Unknown';
    }
}
