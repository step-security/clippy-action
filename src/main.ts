/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023-2024 Noel Towa <cutie@floofy.dev>
 * Copyright (c) 2025 StepSecurity
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { startGroup, setFailed, endGroup, warning, debug, error, info } from '@actions/core';
import { getOctokit, context } from '@actions/github';
import { assertIsError } from '@noelware/utils';
import { getExecOutput } from '@actions/exec';
import { getInputs } from './inputs';
import * as osInfo from './os-info';
import { which } from '@actions/io';
import * as clippy from './clippy';
import axios, { isAxiosError } from 'axios';

async function validateSubscription(): Promise<void> {
    const API_URL = `https://agent.api.stepsecurity.io/v1/github/${process.env.GITHUB_REPOSITORY}/actions/subscription`;

    try {
        await axios.get(API_URL, { timeout: 3000 });
    } catch (err) {
        if (isAxiosError(err) && err.response?.status === 403) {
            error('Subscription is not valid. Reach out to support@stepsecurity.io');
            process.exit(1);
        } else {
            info('Timeout or API not reachable. Continuing to next step.');
        }
    }
}

async function main() {
    await validateSubscription();
    const inputs = getInputs();
    if (inputs === null) process.exit(1);

    startGroup('Check if `cargo` exists');
    let cargoPath: string;
    try {
        cargoPath = await which('cargo', true);
        info(`Found \`cargo\` binary in path [${cargoPath}]`);
    } catch (e) {
        assertIsError(e);
        error("Cargo tool doesn't exist. Please add a step to install a valid Rust toolchain.");

        process.exit(1);
    } finally {
        endGroup();
    }

    startGroup('Collecting rustc information...');
    let version: string;

    {
        const { stdout } = await getExecOutput('rustc', ['--version']);
        version = stdout.slice('rustc '.length);

        endGroup();
    }

    const patch = version.split('.').at(-1);
    if (patch === undefined) {
        process.exit(1);
    }

    const toolchain = patch.endsWith('-nightly') ? 'Nightly' : patch.endsWith('-beta') ? 'Beta' : 'Stable';
    const client = getOctokit(inputs['github-token']);
    const sha = context.sha;
    let canPerformCheckRun = false;
    let id: null | number = null;
    const startedAt = new Date();

    try {
        const { data: newRunData } = await client.request('POST /repos/{owner}/{repo}/check-runs', {
            owner: context.repo.owner,
            repo: context.repo.repo,
            name: `Clippy: Rust ${toolchain} ${version}${
                inputs['working-directory'] !== undefined ? ` in ${inputs['working-directory']}` : ''
            }`,
            head_sha: sha,
            status: 'in_progress',
            started_at: startedAt.toISOString()
        });

        id = newRunData.id;
        canPerformCheckRun = true;
        info(`Created check run with ID [${id}]`);
    } catch (e) {
        warning("clippy-action doesn't have permissions to create Check Runs, disabling!");
        warning(e instanceof Error ? e.message : JSON.stringify(e, null, 4));

        canPerformCheckRun = false;
    }

    const [exitCode, pieces] = await clippy.getClippyOutput(inputs, cargoPath);
    await clippy.renderMessages(pieces);

    const renderer = clippy.kDefaultRenderer;
    const os = osInfo.os();
    const arch = osInfo.arch();
    if (canPerformCheckRun && id !== null) {
        const completed = new Date();
        const { data } = await client.request('PATCH /repos/{owner}/{repo}/check-runs/{check_run_id}', {
            check_run_id: id,
            owner: context.repo.owner,
            repo: context.repo.repo,
            status: 'completed',
            conclusion: exitCode === 0 ? 'success' : 'failure',
            started_at: startedAt.toISOString(),
            completed_at: completed.toISOString(),
            output:
                exitCode === 0
                    ? {
                          title: `Clippy (${toolchain} ~ ${os}/${arch})`,
                          summary: 'Clippy was successful!',
                          text: [
                              `Running \`cargo clippy\` took roughly ~${
                                  completed.getTime() - startedAt.getTime()
                              }ms to complete`,
                              '',
                              `* Working Directory: ${inputs['working-directory'] || 'repository directory'}`
                          ].join('\n'),
                          annotations: renderer.annotations
                              .filter((x) => !!x.file)
                              .map((annotation) => ({
                                  annotation_level:
                                      annotation.level === 'error'
                                          ? ('failure' as const)
                                          : annotation.level === 'warning'
                                            ? ('warning' as const)
                                            : ('notice' as const),
                                  path: annotation.file!,
                                  start_line: annotation.startLine || 0,
                                  end_line: annotation.endLine || 0,
                                  start_column: annotation.startColumn,
                                  end_column: annotation.endColumn,
                                  raw_details: annotation.rendered,
                                  message: annotation.title!
                              }))
                      }
                    : {
                          title: `Clippy (${toolchain} ~ ${os}/${arch})`,
                          summary: 'Clippy failed.',
                          text: [
                              `Running \`cargo clippy\` took roughly ~${
                                  completed.getTime() - startedAt.getTime()
                              }ms to complete`,
                              '',
                              `* Working Directory: ${inputs['working-directory'] || 'repository directory'}`
                          ].join('\n'),
                          annotations: renderer.annotations
                              .filter((x) => !!x.file)
                              .map((annotation) => ({
                                  annotation_level:
                                      annotation.level === 'error'
                                          ? ('failure' as const)
                                          : annotation.level === 'warning'
                                            ? ('warning' as const)
                                            : ('notice' as const),
                                  path: annotation.file!,
                                  start_line: annotation.startLine || 0,
                                  end_line: annotation.endLine || 0,
                                  start_column: annotation.startColumn,
                                  end_column: annotation.endColumn,
                                  raw_details: annotation.rendered,
                                  message: annotation.title!
                              }))
                      }
        });

        debug(JSON.stringify(data));
    }

    info(`Clippy exited with code ${exitCode}`);
    process.exitCode = exitCode;
}

main().catch((ex) => {
    const error = new Error('@augu/clippy-action failed to run', { cause: ex });

    setFailed(error);
    process.exit(1);
});
