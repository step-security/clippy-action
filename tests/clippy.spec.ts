/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023-2024 Noel Towa <cutie@floofy.dev>
 * Copyright (c) 2025 StepSecurity
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getClippyOutput, renderMessages, type Renderer } from '../src/clippy';
import { beforeEach, afterAll, expect, test } from 'bun:test';
import type { AnnotationProperties } from '@actions/core';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { which } from '@actions/io';

const kRenderQueue: Record<
    'error' | 'info' | 'warning',
    Array<[{ properties?: AnnotationProperties; rendered: string }]>
> = {
    warning: [],
    error: [],
    info: []
};

const kMockedRenderer: Renderer = {
    warning(message, properties) {
        kRenderQueue.warning.push([{ rendered: message, properties }]);
    },
    error(message, properties) {
        kRenderQueue.error.push([{ rendered: message, properties }]);
    },
    info(message) {
        kRenderQueue.info.push([{ rendered: message }]);
    }
};

const flushRenderQueue = () => {
    for (const prop of Object.keys(kRenderQueue)) {
        kRenderQueue[prop] = [];
    }
};

test(
    'clippy :: error + warnings',
    async () => {
        const cargo = await which('cargo', true);
        const [exitCode, pieces] = await getClippyOutput(
            {
                'working-directory': resolve(import.meta.dir, '__fixtures__', 'unbased-codebase'),
                'all-features': false,
                'github-token': 'blahblahblah',
                'check-args': [],
                forbid: [],
                allow: [],
                deny: [],
                warn: [],
                args: []
            },
            cargo
        );

        expect(exitCode).toBe(101 /* the code couldn't be compiled correctly */);
        await renderMessages(pieces, kMockedRenderer);
        expect(kRenderQueue.warning.length).toBe(1);
        expect(kRenderQueue.error.length).toBe(1);
        expect(kRenderQueue.info.length).toBe(0);
        expect(kRenderQueue).toMatchSnapshot();

        const warning = kRenderQueue.warning[0];
        expect(warning.length).toBe(1);

        const warningInfo = warning[0];
        expect(warningInfo.rendered).toContain('warning: variable does not');

        const error = kRenderQueue.error[0];
        expect(error.length).toBe(1);

        const errorInfo = error[0];
        expect(errorInfo.rendered).toContain('error: equal expressions as operands to `==`');
    },
    { timeout: 30000 }
);

test('clippy :: no warning or error messages', async () => {
    const cargo = await which('cargo', true);
    const [exitCode, pieces] = await getClippyOutput(
        {
            'working-directory': resolve(import.meta.dir, '__fixtures__', 'no-clippy-error'),
            'all-features': false,
            'github-token': 'blahblahblah',
            'check-args': [],
            forbid: [],
            allow: [],
            deny: ['warnings'],
            warn: [],
            args: []
        },
        cargo
    );

    expect(exitCode).toBe(0);
    await renderMessages(pieces, kMockedRenderer);

    expect(kRenderQueue).toMatchSnapshot();
});
