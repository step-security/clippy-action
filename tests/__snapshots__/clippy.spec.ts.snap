// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`clippy :: error + warnings 1`] = `
{
  "error": [
    [
      {
        "properties": {
          "endColumn": 24,
          "endLine": 5,
          "file": undefined,
          "startColumn": 8,
          "startLine": 5,
          "title": "equal expressions as operands to \`==\`",
        },
        "rendered": 
"error: equal expressions as operands to \`==\`
 --> src/main.rs:5:8
  |
5 |     if *a + 1 == *a + 1 {}
  |        ^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#eq_op
  = note: \`#[deny(clippy::eq_op)]\` on by default

"
,
      },
    ],
  ],
  "info": [],
  "warning": [
    [
      {
        "properties": {
          "endColumn": 14,
          "endLine": 4,
          "file": undefined,
          "startColumn": 9,
          "startLine": 4,
          "title": "variable does not need to be mutable",
        },
        "rendered": 
"warning: variable does not need to be mutable
 --> src/main.rs:4:9
  |
4 |     let mut a = Box::new(1);
  |         ----^
  |         |
  |         help: remove this \`mut\`
  |
  = note: \`#[warn(unused_mut)]\` on by default

"
,
      },
    ],
  ],
}
`;

exports[`clippy :: no warning or error messages 1`] = `
{
  "error": [
    [
      {
        "properties": {
          "endColumn": 24,
          "endLine": 5,
          "file": undefined,
          "startColumn": 8,
          "startLine": 5,
          "title": "equal expressions as operands to \`==\`",
        },
        "rendered": 
"error: equal expressions as operands to \`==\`
 --> src/main.rs:5:8
  |
5 |     if *a + 1 == *a + 1 {}
  |        ^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#eq_op
  = note: \`#[deny(clippy::eq_op)]\` on by default

"
,
      },
    ],
  ],
  "info": [],
  "warning": [
    [
      {
        "properties": {
          "endColumn": 14,
          "endLine": 4,
          "file": undefined,
          "startColumn": 9,
          "startLine": 4,
          "title": "variable does not need to be mutable",
        },
        "rendered": 
"warning: variable does not need to be mutable
 --> src/main.rs:4:9
  |
4 |     let mut a = Box::new(1);
  |         ----^
  |         |
  |         help: remove this \`mut\`
  |
  = note: \`#[warn(unused_mut)]\` on by default

"
,
      },
    ],
  ],
}
`;
